text/0014-network-time.md

```diff
@@ -0,0 +1,683 @@
+
+Example of using Predicted Time:
+
+```csharp
+public class NetworkChat : NetworkBehaviour
+{
+    [ServerRpc(RequireOwnership = false)]
+    SendChatMessageServerRpc(string message, float time)
+    {
+        Debug.Log($"Received message: {message} at time: {time}");
+    }
+
+    public void Update()
+    {
+        if (Input.GetKeyDown(KeyCode.Enter))
+        {
+            SendChatMessageServerRpc("Hello World!", NetworkManager.Singleton.PredictedTime.Time);
+        }
+    }
+}
+```
+*note: this is a stupid example :)*
+
+### Time on the Host/Server
+Since there is 0 latency when making a change to an object on the Host/Server `PredictedTime` will always be the same as `ServerTime` on the Server/Host. When implementing gameplay code using `PredictedTime` is recommended as it allows to potentially re-use the code on a client.
+
+
+### Network Ticks
+
+MLAPI sends, receives and processes data at fixed tick intervals. It is important that gameplay code runs at a the same rate as the fixed tick. This is the case because the state which is synchronized will always be associated with a specific tick. To interpolate state correctly on another peer those tick values are used. This is similar to how you should only move RigidBodies in `FixedUpdate`.
+
+MLAPI exposes multiple APIs to allow developers to do so.
+
+- `NetworkManager.Singleton.NetworkTimeSystem.OnNetworkTick` can be subscribed to to run code every tick.
+- Code running in Update or FixedUpdate could check for the current tick using `NetworkManager.Singleton.NetworkTime.PredictedTime.Tick`.
+- In the future we will expose more convenient ways to run logic during the network tick via a `NetworkFixedUpdate` in `NetworkBehaviour`.
+
+### INetworkTimeProvider
+
+While the built in `NetworkTimeSystem` provides a good starting point for a variety of games, sometimes more control over time is needed. `INetworkTimeProvider` is an interface which allows to override the root calculation of all network time. Note that this is an advanced feature, a faulty implementation of `INetworkTimeProvider` can result in a variety of issues.
+
+Here is an example implementation of a very simple custom `INetworkTimeProvider`:
+```csharp
+public class SampleNetworkTimeProvider : INetworkTimeProvider
+{
+    public bool AdvanceTime(ref NetworkTime predictedTime, ref NetworkTime serverTime, float deltaTime)
+    {
+        serverTime += deltaTime;
+        predictedTime = serverTime + 0.1f;
+        return true;
+    }
+
+    public void InitializeClient(ref NetworkTime predictedTime, ref NetworkTime serverTime)
+    {
+        predictedTime = serverTime + 0.1f;
+    }
+}
+```
+
+This `INetworkTimeProvider` is a very simple provider which increments `ServerTick` and `PredictedTick` at a consistent rate while running the `PredictedTick` 100 ms ahead of the `ServerTick`. This just acts as an example implementation and is not a recommended way to implement a `INetworkTimeProvider`
```

Comment:
"100 ms ahead of the ServerTick"
tick's unit shouldn't be time, it should be an int counting the number of network updates

Reply:
Correct, ticks are ints and not floats. I can see why the explanation here is confusing. ticks are derived from time. Similar to how Unity decides on how many FixedUpdate states it runs in front of each update based on Time.time we do the same thing but with `PredictedTime`.